packages/indexer/native/src/lib.rs

```diff
@@ -398,6 +325,146 @@ declare_types! {
+            };
+            let script = assemble_packed_script(&code_hash, js_hash_type, &args);
+            if script.is_err() {
+                return cx.throw_error(format!("Error assembling script: {:?}", script.unwrap_err()));
+            }
+            let script = script.unwrap();
+
+            let prefix = if cx.argument::<JsNumber>(2)?.value() as u32 == 1 {
+                KeyPrefix::CellTypeScript
+            } else {
+                KeyPrefix::CellLockScript
+            };
+            let args_len = cx.argument::<JsNumber>(3)?.value();
+            if args_len > u32::max_value() as f64 {
+                return cx.throw_error("args length must fit in u32 value!");
+            }
+            let args_len = if args_len <= 0.0 {
+                script.args().len() as u32
+            } else {
+                // when prefix search on args, the args parameter is be shorter than actual args, so need set the args_len manully.
+                args_len as u32
+            };
+            let mut start_key = vec![prefix as u8];
+            start_key.extend_from_slice(script.code_hash().as_slice());
+            start_key.extend_from_slice(script.hash_type().as_slice());
+            start_key.extend_from_slice(&args_len.to_le_bytes());
+            start_key.extend_from_slice(&script.args().raw_data());
+
+            let from_block = cx.argument::<JsValue>(4)?;
+            let to_block = cx.argument::<JsValue>(5)?;
+            let from_block_number_slice = if from_block.is_a::<JsNumber>() {
+                let from_block_number = from_block.downcast::<JsNumber>().or_throw(&mut cx)?.value() as u64;
+                from_block_number.to_be_bytes()
+            } else {
+                0_u64.to_be_bytes()
+            };
+            let to_block_number_slice = if to_block.is_a::<JsNumber>() {
+                // here set to_block_number as toBlock + 1, making the toBlock included in query range.
+                let to_block_number = to_block.downcast::<JsNumber>().or_throw(&mut cx)?.value() as u64 + 1;
+                to_block_number.to_be_bytes()
+            } else {
+                u64::MAX.to_be_bytes()
+            };
+
+            let skip = cx.argument::<JsValue>(6)?;
+            let skip_number = if skip.is_a::<JsNumber>() {
+                skip.downcast::<JsNumber>().or_throw(&mut cx)?.value() as usize
+            } else {
+                0_usize
+            };
+
+            let iter = store.iter(&start_key, IteratorDirection::Forward);
+            if iter.is_err() {
+                return cx.throw_error("Error creating iterator!");
+            }
+
+            let iter = iter.unwrap()
+            .take_while(move |(key, _)| {
+                let block_number_slice = key[key.len() - 16..key.len() - 8].try_into();
+                key.starts_with(&start_key) && to_block_number_slice > block_number_slice.unwrap()
+            })
+            .filter( move |(key, _)| {
```

Comment:
There might be a performance quirk: it really depends on the implementation, but if there are like 2 million blocks, and the toBlock ends at `100`, it might be the case that the code here also needs to loop and filter against all the rest 1999900 blocks, since the iterator might not know that the keys are sorted. I think it's better we manually do an early stop.

Reply:
I considered to perform an early stop like last implementation. But the only possible way I came out is using another rockdb iterator with ` IteratorDirection::Reverse`, so what do you mean an early stop manually here?